charmhelpers/contrib/network/ovs/ovn.py

```diff
@@ -0,0 +1,148 @@
+# Copyright 2019 Canonical Ltd
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#  http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os
+import subprocess
+
+from . import utils
+
+
+def ovn_rundir():
+    """Determine path to OVN sockets.
+
+    Prior to OVN 20.03 the default placement of OVN sockets was together with
+    Open vSwitch, from 20.03 and onwards they are kept in a separate directory.
+
+    :returns: path to OVN rundir
+    :rtype: str
+    """
+    OVN_RUNDIR = '/var/run/ovn'
+    OVS_RUNDIR = '/var/run/openvswitch'
+    for rundir in OVN_RUNDIR, OVS_RUNDIR:
+        if os.path.exists(rundir):
+            return (rundir)
+    # Fall back to new default
+    return OVN_RUNDIR
```

Comment:
I missing something here?  `OVN_RUNDIR` doesn't exist at this point so why return anything?  Why not raise an exception?

Reply:
Prior to Focal and Ussuri UCA OVN was distributed as a part of Open vSwitch and by default placement of config and runtime files was together.  From Focal and Ussuri onwards they are distributed as two separate source packages and default to have configuration and runtime files separate.

A side effect of how the packaging is done pre-Ussuri and post-Ussuri this is a good way to determine which world we live in.

The consuming charm uses this information to figure out in which world we currently live in and may call this function before packages have been installed.  Given we first enabled OVN as a tech preview at Train and the default onward will be the new world order I thought it to be useful to default to the correct value.

Raising an exception here would make the current use and purpose of the function moot and it would basically have to be re-implemented in the consuming charms.

Reply:
Okay, I tend to accept the reasoning.  But would it be better to just check for the non-default path, return that if it exists, and otherwise return the default path.  i.e. only check for the non-default path location?

Reply:
Sure, I'll also revisit the actual need for the helpers, we might also make the library functions default to the current world order and have the consumer optionally override it.

I think this came about before I had proper release selection coded into the charms.